test/common/upstream/zone_aware_load_balancer_fuzz_base.cc

```diff
@@ -0,0 +1,70 @@
+#include "zone_aware_load_balancer_fuzz_base.h"
+
+#include "test/mocks/upstream/host_set.h"
+
+namespace Envoy {
+namespace Upstream {
+
+void ZoneAwareLoadBalancerFuzzBase::initializeASingleHostSet(
+    const test::common::upstream::SetupPriorityLevel& setup_priority_level,
+    const uint8_t priority_level, uint16_t& port) {
+  LoadBalancerFuzzBase::initializeASingleHostSet(setup_priority_level, priority_level, port);
+  // Update local priority set if it exists - will mean load balancer is zone aware and has decided
+  // to construct local priority set
+  if (priority_level == 0 && local_priority_set_.get() != nullptr) {
+    MockHostSet& host_set = *priority_set_.getMockHostSet(priority_level);
+    local_priority_set_->getMockHostSet(0)->hosts_ = host_set.hosts_;
+    local_priority_set_->getMockHostSet(0)->hosts_per_locality_ = host_set.hosts_per_locality_;
+  }
+}
+
+void ZoneAwareLoadBalancerFuzzBase::updateHealthFlagsForAHostSet(
+    const uint64_t host_priority, const uint32_t num_healthy_hosts,
+    const uint32_t num_degraded_hosts, const uint32_t num_excluded_hosts,
+    const std::string random_bytestring) {
+  LoadBalancerFuzzBase::updateHealthFlagsForAHostSet(
+      host_priority, num_healthy_hosts, num_degraded_hosts, num_excluded_hosts, random_bytestring);
+  // Update local priority set if it exists - will mean load balancer is zone aware and has decided
+  // to construct local priority set
+  const uint8_t priority_of_host_set = host_priority % num_priority_levels_;
+  if (priority_of_host_set == 0 && local_priority_set_.get() != nullptr) {
+    MockHostSet& host_set = *priority_set_.getMockHostSet(priority_of_host_set);
+    local_priority_set_->getMockHostSet(0)->healthy_hosts_ = host_set.healthy_hosts_;
```

Comment:
Is this what ` HostSetImpl::updateHosts` does, and if so, does the fuzzer need to ` rebuildLocalityScheduler`?

Reply:
Yeah. Changed this to call into updateHosts, by constructing a updateHostParams struct.

Reply:
Do the other sets need to `rebuildLocalityScheduler`?

Reply:
"Other sets"? I already verified in regard to the base class that the Data Structures would be update due the runCallbacks call at the end of updateHealthFlagsForAHostSet. Harvey had me look into this :)

Reply:
ok! i only bring it up because in addition to running callbacks, that's also run. https://github.com/envoyproxy/envoy/blob/94eec5b84c42b0c42017681271d955525c9533e8/source/common/upstream/upstream_impl.cc#L418

Reply:
Interesting, let me look at what I looked at for Harvey's comment. I know it pipes in things to callbacks, but perhaps this doesn't get called like you mentioned. I'll take a look around.

Reply:
Ah - great catch. The data structures that get updated previously were those here: https://github.com/envoyproxy/envoy/blob/5875f23b60aeeb49e038596d8de36446c6728f5d/source/common/upstream/load_balancer_impl.cc#L116. Thanks.

Reply:
Unit tests don't have any calls to updateHosts or anything after setting weights, but some interesting callback logic. I'll investigate (I bet updateHosts() and rebuildLocalityScheduler() is somehow being called in that callback)

Reply:
![Screenshot 2020-10-29 at 2 35 33 PM](https://user-images.githubusercontent.com/39203661/97617445-29447900-19f4-11eb-8089-6be52730483f.png)

Reply:
Hey Asra, ran coverage on unit tests, and it doesn't hit the logic here. Running on benchmark and simulation now.

Reply:
Running now on my fuzz test to see if it gets to the same if clause :)

Reply:
To document, this current commit is wrong. Rewriting it now. What's currently checked in never called priority set update logic.

Reply:
![Screenshot 2020-10-29 at 4 04 02 PM](https://user-images.githubusercontent.com/39203661/97626554-66166d00-1a00-11eb-976d-cccff820abeb.png)

Reply:
Ok, rewrote my fuzzer, coverage does match up now :).